mmseg/models/losses/hausdorff_distance_loss.py

```diff
@@ -0,0 +1,141 @@
+(Apache-2.0 license)"""
+import numpy as np
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from scipy.ndimage import distance_transform_edt
+
+from ..builder import LOSSES
+from .utils import get_class_weight, weighted_loss
+
+
+def compute_dtm(mask_arr, output_shape, normalized=False):
+    """compute the distance transform map of foreground in binary mask.
+
+    Args:
+        mask_arr (array): segmentation mask, shape=(batch_size, channel, x, y)
+        output_shape (tuple): predict mask shape after F.Softmax(dim=1),
+                              shape=(batch_size, 2, channel, x, y)
+        normalized (bool): whether compute the dtm by normalized
+
+    Returns:
+        fg_dtm (array)：the foreground Distance Map (SDM), shape=out_shape
+            dtm(x) = 0;     x out of segmentation or x in segmentation boundary
+                   = inf|x - y|;    x in segmentation
+                                    if normalized is True, dtm(x) to [0, 1]
+    """
+    fg_dtm = np.zeros(output_shape)
+
+    for b in range(output_shape[0]):
+        for c in range(1, output_shape[1]):
+            positive_mask = mask_arr[b].astype(bool)
+            if positive_mask.any():
+                positive_distance = distance_transform_edt(positive_mask)
+                fg_dtm[b][c] = positive_distance/np.max(positive_distance) \
+                    if normalized else positive_mask
+    return fg_dtm
+
+
+@weighted_loss
+def binary_hausdorff_distance_loss(pred, target, valid_mask):
+    assert pred.shape[0] == target.shape[0]
+
+    with torch.no_grad:
+        gt_dtm_npy = compute_dtm((target * valid_mask).cpu.numpy(), pred.shape)
+        gt_dtm = torch.from_numpy(gt_dtm_npy).float().cuda(pred.device.index)
+        pred_dtm_npy = compute_dtm(
+            (pred * valid_mask)[:, 1, ...].cpu.numpy() > 0.5, pred.shape)
+        pred_dtm = torch.from_numpy(pred_dtm_npy).float().cuda(
+            pred.device.index)
+
+    # compute hausdorff distance loss for binary segmentation
+    delta_s = (pred[:, 1, ...] - target.float())**2
+    dtm = pred_dtm[:, 1, ...]**2 + gt_dtm[:, 1, ...]**2
+    hd_loss = torch.einsum('bcxy,bcxy->bcxy', delta_s, dtm).mean()
+
+    return hd_loss
+
+
+@weighted_loss
+def hausdorff_distance_loss(pred,
+                            target,
+                            valid_mask,
+                            class_weight=None,
+                            ignore_index=255):
+    assert pred.shape[0] == target.shape[0]
+    total_loss = 0
+    num_classes = pred.shape[1]
+    for i in range(num_classes):
+        if i != ignore_index:
+            hd_loss = binary_hausdorff_distance_loss(
+                pred[:, i], target[..., i], valid_mask=valid_mask)
+            if class_weight is not None:
+                hd_loss *= class_weight[i]
+            total_loss += hd_loss
+    return total_loss / num_classes
+
+
+@LOSSES.register_module()
+class HausdorffDistanceLoss(nn.Module):
```

Comment:
Please add unit test, followed this PR: https://github.com/open-mmlab/mmsegmentation/pull/2000.

Its unit test file is [here](https://github.com/open-mmlab/mmsegmentation/blob/dev-1.x/tests/test_models/test_losses/test_tversky_loss.py).

Reply:
OK, thanks for your kindly reminder